relay-server/src/utils/memory.rs

```diff
@@ -0,0 +1,269 @@
+struct Inner {
+    memory: ArcSwap<Memory>,
+    last_update: AtomicU64,
+    reference_time: Instant,
+    system: Mutex<System>,
+}
+
+/// Wrapper around [`Inner`] which hides the [`Arc`] and exposes utils method to make working with
+/// [`MemoryStat`] as opaque as possible.
+#[derive(Clone)]
+pub struct MemoryStat(Arc<Inner>);
+
+impl MemoryStat {
+    /// Creates an instance of [`MemoryStat`] and obtains the current memory readings from
+    /// [`System`].
+    pub fn new() -> Self {
+        // sysinfo docs suggest to use a single instance of `System` across the program.
+        let mut system = System::new();
+        Self(Arc::new(Inner {
+            memory: ArcSwap::from(Arc::new(Self::refresh_memory(&mut system))),
+            last_update: AtomicU64::new(0),
+            reference_time: Instant::now(),
+            system: Mutex::new(system),
+        }))
+    }
+
+    /// Returns a copy of the most up to date [`Memory`] data.
+    pub fn memory(&self) -> Memory {
+        self.try_update();
+        **self.0.memory.load()
+    }
+
+    /// Builds a [`MemoryStatConfig`] which holds a reference to the supplied [`Config`] and the
+    /// current [`MemoryStat`].
+    pub fn with_config(&self, config: Arc<Config>) -> MemoryStatConfig {
+        MemoryStatConfig {
+            memory_stat: self.clone(),
+            config,
+        }
+    }
+
+    /// Refreshes the memory readings.
+    fn refresh_memory(system: &mut System) -> Memory {
+        system.refresh_memory_specifics(MemoryRefreshKind::new().with_ram());
+        let memory = match system.cgroup_limits() {
+            Some(cgroup) => Memory {
+                used: cgroup.rss,
+                total: cgroup.total_memory,
+            },
+            None => Memory {
+                used: system.used_memory(),
+                total: system.total_memory(),
+            },
+        };
+
+        metric!(gauge(RelayGauges::SystemMemoryUsed) = memory.used);
+        metric!(gauge(RelayGauges::SystemMemoryTotal) = memory.total);
+
+        memory
+    }
+
+    /// Updates the memory readings if at least [`UPDATE_TIME_THRESHOLD_MS`] has passed.
+    fn try_update(&self) {
+        let last_update = self.0.last_update.load(Ordering::Relaxed);
+        let elapsed_time = self.0.reference_time.elapsed().as_millis() as u64;
+
+        if elapsed_time - last_update < UPDATE_TIME_THRESHOLD_MS {
+            return;
+        }
+
+        let Ok(_) = self.0.last_update.compare_exchange_weak(
+            last_update,
+            elapsed_time,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        ) else {
+            return;
+        };
+
+        let mut system = self
+            .0
+            .system
+            .lock()
+            .unwrap_or_else(|system| system.into_inner());
+
+        let updated_memory = Self::refresh_memory(&mut system);
+        self.0.memory.store(Arc::new(updated_memory));
+    }
```

Comment:
Not sure if this would simplify things, but instead of calling `compare_exchange_weak` we could call `try_lock()` on `system`. If we fail to acquire the lock, another thread is already updating the stats. If we succeed in acquiring the lock, we can safely update `memory` and `last_update` from within the critical section.

Reply:
Yes, this is also a good idea. I would be curious to know about the performance implications of the `try_lock` versus a `cas` operation. Locks might use `cas` inside so maybe the difference is not that big.